src/ipc/ipc4/mtrace.c

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Kai Vehmanen <kai.vehmanen@linux.intel.com>
+
+#include <stdint.h>
+#include <sof/lib/uuid.h>
+#include <sof/ipc/common.h>
+#include <ipc4/base_fw.h>
+#include <ipc4/error_status.h>
+
+#ifdef CONFIG_LOG_BACKEND_ADSP_MTRACE
+
+#undef panic // SOF headers may redefine panic
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log.h>
+
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/schedule.h>
+#include <sof/schedule/task.h>
+
+LOG_MODULE_REGISTER(mtrace, CONFIG_SOF_LOG_LEVEL);
+
+/**
+ * If mtrace log buffer has less space than
+ * the threshold, notify host with BUFFER_STATUS IPC4.
+ */
+#define NOTIFY_BUFFER_STATUS_THRESHOLD 2048
+
+/**
+ * Default aging timer value. This time defines the maximum time to
+ * block IPC BUFFER STATUS notify messages. The notification is sent
+ * either due to amount of data available in the buffer, or by timeout
+ * in case logs are sent at slow rate.
+ */
+#define IPC4_MTRACE_NOTIFY_AGING_TIMER_MS 1000
+
+/**
+ * Smallest possible aging timer value
+ */
+#define IPC4_MTRACE_AGING_TIMER_MIN_MS 100
+
+/* bb2aa22e-1ab6-4650-8501-6e67fcc04f4e */
+DECLARE_SOF_UUID("mtrace-task", mtrace_task_uuid, 0xbb2aa22e, 0x1ab6, 0x4650,
+		 0x85, 0x01, 0x6e, 0x67, 0xfc, 0xc0, 0x4f, 0x4e);
+
+static uint64_t mtrace_notify_last_sent;
+
+static uint32_t mtrace_bytes_pending;
+
+static uint32_t mtrace_aging_timer = IPC4_MTRACE_NOTIFY_AGING_TIMER_MS;
+
+static struct k_thread mtrace_thread;
+
+#define MTRACE_STACK_SIZE	4096
+#define MTRACE_IPC_CORE	PLATFORM_PRIMARY_CORE_ID
+#define MTRACE_THREAD_PRIO      -1
+#define MTRACE_THREAD_NAME      "mtrace"
+
+K_KERNEL_STACK_DEFINE(mtrace_stack, MTRACE_STACK_SIZE);
+
+static void mtrace_log_hook(size_t written, size_t space_left)
+{
+	if (arch_proc_id() == MTRACE_IPC_CORE &&
+	    (space_left < NOTIFY_BUFFER_STATUS_THRESHOLD ||
+	     k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer)) {
+		ipc_send_buffer_status_notify();
+		mtrace_notify_last_sent = k_uptime_get();
+		mtrace_bytes_pending = 0;
+
+	} else {
+		mtrace_bytes_pending += written;
+	}
+}
+
+static void mtrace_thread_fn(void *p1, void *p2, void *p3)
+{
+	while (1) {
+		if (k_uptime_delta(&mtrace_notify_last_sent) >= mtrace_aging_timer &&
```

Comment:
if using `k_wakeup()` we'd have to drop this timeout check

Reply:
I think we need always a timer. The implementation could be improved by reprogramming the timer whenever new data is added to the buffer (if data is added at a constant fast enough pace, mtrace_thread_fn doesn't need to run ever), but this does not seem to be warranted and the simpler code does the job.

Reply:
Note that this is recapitulating very similar logic in use in the Zephyr logging subsystem when CONFIG_LOG_DEFERRED=y.   You could almost certainly freeload on top of that instead (IIRC your "process" callback happens in the log thread on a timer).  @nordic-krch would be the expert.

Threads are very expensive if all you want is a regular low-priority function called.  Another option would be to use a system work queue callback that just reschedules itself.

Reply:
@andyross Thanks! I now took a tour of the log-deferred code and yeah, I'm doing a double here. Hmm. I did look at system work queue, but I couldn't figure out how to ensure the callback would be pinned on the specific core (and it wouldn't really make sense for the system wq thread).  

Hmm. So the k_thread is a now fairly costly workaround for inability to use the IPC stack from multiple cores.

On cache coherent systems, it'd be obvious to just change the IPC code, but in this case there's always a cost. Starts to seem like a per-cpu wq would be very practical on these systems. We can of course do this at app (SOF) level the SOF LL scheduler in practise is just the thing, it creates core-pinned threads and provides an API to get work done on these pinned threads with domain specific scheduling order.

Maybe I need to switch back to earlier version and use the SOF pinned threads.

Reply:
Oh sorry, I didn't realize this was part of the same core 0 issue. :)

But actually the same ideas hold here too: that logging thread could absolutely be pinned to core 0 if that's what the application wants.  You could probably do it with a quick hack out of SOF code, but a kconfig like "CONFIG_LOG_THREAD_CPUMASK" would be easy to do and easy to support.

Reply:
> I think we need always a timer.

sorry, I wasn't proposing to drop the thread, on the contrary - I was proposing to wake it up to send the notification from its context on core 0 always instead of potentially notifying from any other core where logging is called. Then you'd only send notifications from the thread and the hook would just wake it up.

Reply:
> Threads are very expensive

@andyross hmm, it looks pretty with a thread though :-) Well, if it's really too expensive, a work would be good too, if it can be pinned to core 0... Although do we need that? Would anything break if we allow that work to run on any core?

Reply:
Heh, "too expensive" is an app decision.  But this thread in particular is being granted a 4k stack, vs. the (I think) 24 bytes required for a workq item.  (I'm sure it doesn't need that much, though getting xtensa calls onto stacks smaller than 512 bytes is very difficult; and in any case when we go to an MMU environment in the future 4k will be the mapping granularity anyway).

If you do want to pin the system work queue though, you can find the thread in the system work queue object, which is a public symbol in kernel.h, something like:

```
/* Do this early, before any items get added */
k_thread_cpu_pin(&k_sys_work_q.thread, 0);
```

But @lgirdwood is right, this isn't a -1 and there's no point in holding things up for what amount to

Reply:
@andyross Pinning the sys_work_q is tempting, but seems a bit evasive. We might have good reasons to schedule sys_work_q stuff on any available core.

I now took yet another approach and implemented the worker logic with SOF's EDF scheduling API. This basicly creates a k_thread pinned on a CPU anyways, so why not piggy back on this.